include/Dialect/BGV/IR/BGVOps.td

```diff
@@ -0,0 +1,142 @@
+include "BGVTypes.td"
+
+include "mlir/IR/OpBase.td"
+include "mlir/Interfaces/InferTypeOpInterface.td"
+
+// TODO(https://github.com/google/heir/issues/101): Add verifiers for BGV ops.
+
+//===----------------------------------------------------------------------===//
+// BGV Operation definitions.
+//===----------------------------------------------------------------------===//
+
+class BGV_Op<string mnemonic, list<Trait> traits = []> :
+        Op<BGV_Dialect, mnemonic, traits> {
+
+  // See https://mlir.llvm.org/docs/DefiningDialects/Operations/#declarative-assembly-format
+  // TODO(b/291601096): Simplify this format by adding type constraints.
+  let assemblyFormat = [{
+    `(` operands `)` attr-dict `:`  `(` type(operands) `)` `->` type(results)
+  }];
+  let cppNamespace = "::mlir::heir::bgv";
+}
+
+//===----------------------------------------------------------------------===//
+// Homomorphic operations.
+//===----------------------------------------------------------------------===//
+
+// TODO(https://github.com/google/heir/issues/100): Add plaintext-ciphertext operations.
+
+def BGV_AddOp : BGV_Op<"add", [Commutative, SameOperandsAndResultType]> {
+  let summary = "Addition operation between ciphertexts.";
+
+  let arguments = (ins
+    Ciphertext:$x,
+    Ciphertext:$y
+  );
+
+  let results = (outs
+    Ciphertext:$output
+  );
+
+  let assemblyFormat = "`(` operands `)` attr-dict `:` type($output)" ;
+}
+
+def BGV_SubOp : BGV_Op<"sub", [SameOperandsAndResultType]> {
+  let summary = "Subtraction operation between ciphertexts.";
+
+  let arguments = (ins
+    Ciphertext:$x,
+    Ciphertext:$y
+  );
+
+  let results = (outs
+    Ciphertext:$output
+  );
+
+  let assemblyFormat = "`(` operands `)` attr-dict `:` type($output)" ;
+}
+
+def BGV_MulOp : BGV_Op<"mul", [Commutative]> {
+  let summary = "Multiplication operation between ciphertexts.";
+
+  let arguments = (ins
+    Ciphertext:$x,
+    Ciphertext:$y
+  );
+
+  let results = (outs
+    Ciphertext:$output
+  );
+
+  let assemblyFormat = "`(` operands `)` attr-dict `:` type($x)`,` type($y) `->` type($output)" ;
+}
+
+//===----------------------------------------------------------------------===//
+// Ring operations.
+//===----------------------------------------------------------------------===//
+
+def BGV_Rotate : BGV_Op<"rotate"> {
+  let summary = "Rotate the coefficients of the ciphertext using the Galois Key.";
+
+  let arguments = (ins
+    Ciphertext:$x,
+    I64Attr:$offset
+  );
+
+  let results = (outs
+    Ciphertext:$output
+  );
+}
+
+def BGV_Negate : BGV_Op<"negate", [SameOperandsAndResultType]> {
+  let summary = "Negate the coefficients of the ciphertext.";
+
+  let arguments = (ins
+    Ciphertext:$x
+  );
+
+  let results = (outs
+    Ciphertext:$output
+  );
+}
+
+def BGV_Relinearize : BGV_Op<"relinearize"> {
+  let summary = "Relineraize the ciphertext.";
+
+  let arguments = (ins
+    Ciphertext:$x,
+    DenseI32ArrayAttr:$from_basis,
```

Comment:
I feel like the semantics of the basis could use a bit more documentation, at least to explain how `(1, s, s^2)` corresponds to the attribute `[0, 1, 2]` while a `(1, s^2)` would correspond to `[0, 2]`

Reply:
added a description body to this op